pkg/util/ring/adaptive_chan.go

```diff
@@ -0,0 +1,310 @@
+			dequeueChan := internalRecv
+			if stats.Len == 0 {
+				dequeueChan = nil
+			}
+
+			select {
+			case v, ok := <-internalSend: // blocks until something is queued
+				if !ok {
+					// in was closed, so if we leave it like that the next
+					// iteration will keep receiving zero values with ok=false
+					// without any blocking. So we set in to nil, so that
+					// the next iteration the select will block indefinitely on
+					// this branch of the select and leave only the dequeing
+					// branch active until all items have been dequeued
+					internalSend = nil
+				} else {
+					q.Enqueue(v)
+				}
+
+			case dequeueChan <- q.Peek(): // blocks if nothing to dequeue
+				// we don't want to call Dequeue in the `case` above since that
+				// would consume the item and it would be lost if the queueing
+				// branch was selected, so instead we Peek in the `case` and we
+				// do the actual dequeueing here once we know this branch was
+				// selected
+				q.Dequeue()
+
+			case statsChan <- stats:
+				// stats reading
+				stats.StatsRead++
+
+			case cmd, ok := <-cmdChan:
+				if !ok {
+					// AdaptiveChanController was closed. Set cmdChan to nil so
+					// this branch blocks in the next iteration
+					cmdChan = nil
+					continue
+				}
+
+				// execute a command on the internal *Ring
+				switch cmd.acCmdType {
+				case acCmdMin:
+					q.Min = cmd.intValue
+					stats.Min = cmd.intValue
+				case acCmdMax:
+					q.Max = cmd.intValue
+					stats.Max = cmd.intValue
+				case acCmdGrow:
+					q.Grow(cmd.intValue)
+				case acCmdShrink:
+					q.Shrink(cmd.intValue)
+				case acCmdClear:
+					q.Clear()
+				}
+			}
+		}
+	}()
+
+	return internalSend, internalRecv, ctrl
+}
+
+type acCmdType uint8
+
+const (
+	acCmdMin = iota
+	acCmdMax
+	acCmdClear
+	acCmdGrow
+	acCmdShrink
+)
+
+type acCmd struct {
+	acCmdType
+	intValue int
+}
+
+// AdaptiveChanController provides access to an AdaptiveChan's internal *Ring.
+type AdaptiveChanController struct {
+	statsChan <-chan AdaptiveChanStats
+	cmdChan   chan<- acCmd
+	cmdChanMu sync.Mutex
+	closed    uint32
+}
+
+// Close releases resources associated with this controller. After calling this
+// method, all other methods will return ErrAdaptiveChanControllerClosed. It is
+// idempotent. This doesn't affect the queue itself, but rather prevents further
+// administrative tasks to be performed through the AdaptiveChanController.
+func (r *AdaptiveChanController) Close() {
+	r.cmdChanMu.Lock()
+	defer r.cmdChanMu.Unlock()
+
+	if r.cmdChan != nil {
+		close(r.cmdChan)
+		r.cmdChan = nil
+		atomic.StoreUint32(&r.closed, 1)
+	}
+}
+
+func (r *AdaptiveChanController) isClosed() bool {
+	return atomic.LoadUint32(&r.closed) != 0
+}
+
+// Min sets the value of Min in the internal *Ring.
+func (r *AdaptiveChanController) Min(ctx context.Context, n int) error {
+	if r.isClosed() {
+		return ErrAdaptiveChanControllerClosed
+	}
+
+	r.cmdChanMu.Lock()
+	defer r.cmdChanMu.Unlock()
+
```

Comment:
do we need r.closed vs just checking whether r.cmdChan is nil after getting the mutex but before trying to send (which we already do in sendOrErr)?

I see that we call isClosed without acquiring the cmdChanMu in WriteStats, do we actually need that check?

we could also reduce repetition by implementing a sendCmd(ctx, acCmd) method on AdaptiveChanController and let that handle the mutex and closed check

Reply:
Fixed!